plugins/optimization-detective/optimization.php

```diff
@@ -32,16 +32,19 @@
  */
 function od_buffer_output( string $passthrough ): string {
 	ob_start(
-		static function ( string $output ): string {
-			/**
-			 * Filters the template output buffer prior to sending to the client.
-			 *
-			 * @since 0.1.0
-			 *
-			 * @param string $output Output buffer.
-			 * @return string Filtered output buffer.
-			 */
-			return (string) apply_filters( 'od_template_output_buffer', $output );
+		static function ( string $output, ?int $phase ): string {
+			if ( ( $phase & PHP_OUTPUT_HANDLER_FINAL ) > 0 ) {
```

Comment:
A few questions so that I understand this: 
* This is only relevant for when someone calls `ob_clean()`? Or for other things as well?
* What's the purpose of `$phase` and what's the purpose of `PHP_OUTPUT_HANDLER_FINAL` in this? Why are both checks needed?

Reply:
> * This is only relevant for when someone calls `ob_clean()`? Or for other things as well?

That's a good point. There's also the consideration of `ob_flush()`. Currently if this gets called then the resulting buffer passed into the filter is not complete. I've improved this in f218dc833f9bb02b0001e9a765fbebb23ddd530a by making it so that the buffer is cleanable but not flushable. 

> * What's the purpose of `$phase` and what's the purpose of `PHP_OUTPUT_HANDLER_FINAL` in this? Why are both checks needed?

The `$phase` provides information about [why the handler was invoked](https://www.php.net/manual/en/outcontrol.flags-passed-to-output-handlers.php). Note that bitwise operations are being used here (`&` not `&&`), so it is checking if the `$phase` contains the final flag.

Reply:
@westonruter Thanks, I had indeed missed the bitwise &, so this is only one check, makes sense.

Reading your code comments around the custom flag handling, I'm trying to understand: Does this mean that calling `ob_flush()` anywhere while this output buffer is active will trigger an error? If so, I think we need to review whether/how that function is commonly used by plugins, mostly caching plugins of course. While it seems necessary for our functionality to be reliable, it also seems risky in terms of cross-plugin compatibility, like you're already indicating in your TODO.

Reply:
It won't cause an error, no. It just results in `ob_flush()` returning false, which is included in the unit tests.

I'm addressing the TODO in another PR as it's not really related to output buffering.

Reply:
Ok, but what about breakage by flushing not being possible in a plugin where that might be intended? That's where I think we need to take a look at the ecosystem to see if that's a real or just a theoretical problem.

Reply:
That would be a situation where a plugin is attempting to flush the output buffer for a buffer it didn't open, which would be poor form. I did look over the [instances of `ob_flush()` in WPDirectory](https://wpdirectory.net/search/01J4MKNHR6MSNT4VMEZ1AJN19M) and I didn't see any obvious examples where `ob_flush()` was being used in frontend contexts without also having called `ob_start()`: so they'd only be flushing their own buffer. So I think we're fine here. We'll need to do more testing with actual sites to discover if there is a compatibility problem.